### PR TITLE
Core SDK v1.0.0

### PR DESCRIPTION
**Pull request by zatlite**
# Summary

- Expose fewer public methods to make it easier for us to maintain the SDK.
- Introduced a [single entry point](https://docs.x.immutable.com/sdk-references/core-sdk-ts/1.0.0-beta.2/classes/immutablex.immutablex) for the SDK to improve discoverability 
- Clear response and error types, no more Axios wrappers
- Simplified complex types required for creating trades, orders and transfers, no more SignableToken in the interface
- Make the deposit method on SDK consistent with the API and industry norms. 
- SDK now handles L1 signature-protected APIs; no need to generate imx-signature separately for authentication headers
- Removed wallet-specific logic
- Removed the `ethereumjs-wallet` dependency.

[DX-1018]

# Why the changes

# Things worth calling out

[DX-1018]: https://immutable.atlassian.net/browse/DX-1018?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,21 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
+
+### Changed
+
+- Significant changes to public interface: ImmutableX
```
Comment (Kastet) [inline CHANGELOG.md:12]:
```suggestion
- Significant changes to the public interface. Introduced ImmutableX class as a single entry point for the SDK to improve discoverability
```

Comment (Kastet) [inline CHANGELOG.md:12]: This line can probably be removed now

Comment (zatlite) [reply CHANGELOG.md]: [627d900](https://github.com/immutable/imx-core-sdk/pull/185/commits/627d9004b0fdcff0cad0ca82adc9e0a9bf8bbb66)

File: src/types/index.ts [hunk 2]
```diff
@@ -1,48 +1,5 @@
...
-export type UnsignedMintRequest = Omit<MintRequest, 'auth_signature'>;
-export type UnsignedTransferRequest = GetSignableTransferRequestV1;
-export type UnsignedBatchNftTransferRequest = GetSignableTransferRequest;
-export type UnsignedBurnRequest = GetSignableBurnRequest;
-
-export * from './deposit';
-export * from './withdrawal';
-export * from './token';
-export * from './signable-withdrawal';
+export * from './generic';
```
Comment (Kastet) [inline src/types/index.ts:1]: file name suggestion: `publicInterfaces` instead of `generic`. thoughts @matthewmuscat @zatlite ?

**Review by matthewmuscat — APPROVED** [review]

**Review by matthewmuscat — CHANGES_REQUESTED** [review]
Ignore my approval ^

File: src/types/index.ts (hunk 2 above)
Comment (zatlite) [reply src/types/index.ts:1]: @Kastet How about we split the generic file into `signerInterfaces` and `configInterfaces`?

Comment (Kastet) [reply src/types/index.ts:1]: or config interfaces can go into `config.ts` and this file can be renamed to `signers.ts`?

Comment (zatlite) [reply src/types/index.ts]: [69dd24a](https://github.com/immutable/imx-core-sdk/pull/185/commits/69dd24a1f887bea007bc7a3990422774e93afbdf)

**Review by Kastet — APPROVED** [review]
LGTM!

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,6 +5,20 @@ All notable changes to this project will be documented in this file.
...
+- Expose fewer public methods to make it easier for us to maintain the SDK.
+- Introduced a [single entry point](https://github.com/immutable/imx-core-sdk/blob/34629c169f2c0fc6e38e73ec75a814f36b680620/src/ImmutableX.ts#L64) for the SDK to improve discoverability
+- Clear response and error types, no more Axios wrappers
+- Simplified complex types required for creating trades, orders and transfers, no more SignableToken in the interface
+- Make the deposit method on SDK consistent with the API and industry norms.
+- SDK now handles L1 signature-protected APIs; no need to generate imx-signature separately for authentication headers
+- Removed wallet-specific logic
+- Removed the `ethereumjs-wallet` dependency.
+
+
```
Comment (matthewmuscat) [inline CHANGELOG.md:21]: Do we need two empty lines? One should be fine.

File: src/ImmutableX.ts [hunk 4]
```diff
@@ -0,0 +1,842 @@
...
+      .getDeposit(request)
+      .then(res => res.data)
+      .catch(err => {
+        throw formatError(err);
+      });
+  }
+
+  /**
+   * Get a list of Deposits
+   * @param request - the request object containing the parameters to be provided in the API request
```
Comment (matthewmuscat) [inline src/ImmutableX.ts:131]: `request` is optional? Should we specify this? (not blocking)

File: src/ImmutableX.ts [hunk 5]
```diff
@@ -0,0 +1,842 @@
...
+      .getAsset(request)
+      .then(res => res.data)
+      .catch(err => {
+        throw formatError(err);
+      });
+  }
+
+  /**
+   * Get a list of Assets
+   * @param request - the request object containing the parameters to be provided in the API request
```
Comment (matthewmuscat) [inline src/ImmutableX.ts:200]: another one here, looks like all the `list` methods are optional param.

File: src/exportContracts.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+export {
+  Core__factory as Core,
```
Comment (matthewmuscat) [inline src/exportContracts.ts:2]: @Kastet @zatlite Did we manage to figure out why the typedoc still generates these as the former names?

File: src/utils/convertToSignableToken.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import { SignableToken } from '../api';
+import { TokenAmount } from '../types/tokens';
+
+/**
+ * Helper method to convert token type to a SignableToken type
+ * @param token - the token type to convert to a SignableToken type
+ * @returns the converted SignableToken
+ */
+
```
Comment (matthewmuscat) [inline src/utils/convertToSignableToken.ts:9]: Does this annotation still work with this new line gap? Can we remove this?

File: src/utils/formatError.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import axios from 'axios';
+import { APIError } from '../api';
+import { IMXError } from '../types/errors';
+
+/**
+ * [Formats an error in the IMXError shape](https://axios-http.com/docs/handling_errors)
+ * @param error - The Error object thrown by the request
+ * @returns IMXError
+ */
+
```
Comment (matthewmuscat) [inline src/utils/formatError.ts:10]: Same here, can we remove this line?

File: src/utils/formatError.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  if (axios.isAxiosError(error) && error.response) {
+    const apiError: APIError = error.response.data;
+    return new IMXError({
+      code: apiError.code,
+      details: apiError.details,
+      message: apiError.message,
+    });
+  }
+  return new IMXError({
+    code: 'unknown_error_code',
```
Comment (matthewmuscat) [inline src/utils/formatError.ts:21]: Would it be worth adding this to our docs, [here?](https://docs.x.immutable.com/docs/error-codes)

File: src/workflows/deposit/depositERC20.ts [hunk 10]
```diff
@@ -88,12 +89,12 @@ export async function depositERC20Workflow(
     token_address: deposit.tokenAddress,
   };
 
-  const amount = parseUnits(deposit.amount, BigNumber.from(decimals));
+  const amount = parseUnits(deposit.amount, 0); // 0 to always use undecimilised value
```
Comment (matthewmuscat) [inline src/workflows/deposit/depositERC20.ts:92]:
```suggestion
  const amount = parseUnits(deposit.amount, 0); // 0 to always use undecimalized value
```

File: src/workflows/registration.ts [hunk 11]
```diff
@@ -54,14 +42,25 @@ export async function registerOffchainWorkflow({
     },
   });
 
-  return;
+  return registeredUser.data;
```
Comment (matthewmuscat) [inline src/workflows/registration.ts:45]: We decided that this should be fixed in the upstream api-level as it always returns `{"tx_hash": ""}`. Have we followed up / made a ticket to action this? If not, would be good as we will likely not revisit this for a while.

File: src/workflows/registration.ts [hunk 12]
```diff
@@ -54,14 +42,25 @@ export async function registerOffchainWorkflow({
...
 ): Promise<boolean> {
-  return await contract.isRegistered(starkPublicKey);
+  try {
+    return await contract.isRegistered(starkPublicKey);
+  } catch (ex) {
+    if ((ex as IsRegisteredCheckError).reason === 'USER_UNREGISTERED') {
+      return false;
+    }
+    throw ex;
+  }
```
Comment (matthewmuscat) [inline src/workflows/registration.ts:63]: cleaner suggestion:

```suggestion
  try {
    return await contract.isRegistered(starkPublicKey);
  } catch (err) {
    if (err.reason === 'USER_UNREGISTERED') return false;
    throw ex;
  }
```

File: src/workflows/transfers.ts [hunk 13]
```diff
@@ -69,15 +74,29 @@ export async function transfersWorkflow({
...
-  l1Signer,
-  l2Signer,
+  ethSigner,
+  starkSigner,
   request,
   transfersApi,
 }: BatchTransfersWorkflowParams): Promise<CreateTransferResponse> {
+  const ethAddress = await ethSigner.getAddress();
+
+  const signableRequests = request.map(x => {
```
Comment (matthewmuscat) [inline src/workflows/transfers.ts:84]: What is `x`? Can we make this name clearer or strictly type the param?

File: src/workflows/withdrawal/completeERC721Withdrawal.ts [hunk 14]
```diff
@@ -263,7 +255,7 @@ export async function completeERC721WithdrawalWorkflow(
       ),
     )
     .catch(error => {
-      if (error.response.status === 404) {
+      if (error.response?.status === 404) {
         console.log(error.response); // token is already minted on L1
```
Comment (matthewmuscat) [inline src/workflows/withdrawal/completeERC721Withdrawal.ts:259]: Do we want this log?

**Review by matthewmuscat — COMMENTED** [review]
Looks great! A few things..

File: src/utils/convertToSignableToken.ts (hunk 7 above)
Comment (zatlite) [reply src/utils/convertToSignableToken.ts:9]: These methods are not exposed so you wouldn't find it in typedoc but yes it does work (see intellisense) with the line gap. I'll remove the gaps though.

File: src/utils/formatError.ts (hunk 9 above)
Comment (zatlite) [reply src/utils/formatError.ts:21]: That page is for API error codes and this unknown error would include errors unrelated to API for example when you do workflow steps. So I don't think it is the right place for it.

File: src/ImmutableX.ts (hunk 4 above)
Comment (zatlite) [reply src/ImmutableX.ts:131]: typedoc already marks it as optional but I will update the description from `- the request object` to `- optional request object`

image

Comment (matthewmuscat) [reply src/ImmutableX.ts:131]: that's nice, how about for intellisense?

File: src/utils/formatError.ts (hunk 9 above)
Comment (matthewmuscat) [reply src/utils/formatError.ts:21]: good point. I think there's talk of turning these into enums as a source of truth, so we can use this then.

File: CHANGELOG.md (hunk 3 above)
Comment (zatlite) [reply CHANGELOG.md]: [05dab29](https://github.com/immutable/imx-core-sdk/pull/185/commits/05dab29a61dda6416ec0080c8c45a19130e36a68)

File: src/ImmutableX.ts (hunk 4 above)
Comment (zatlite) [reply src/ImmutableX.ts]: [05dab29](https://github.com/immutable/imx-core-sdk/pull/185/commits/05dab29a61dda6416ec0080c8c45a19130e36a68)

File: src/ImmutableX.ts (hunk 5 above)
Comment (zatlite) [reply src/ImmutableX.ts]: [05dab29](https://github.com/immutable/imx-core-sdk/pull/185/commits/05dab29a61dda6416ec0080c8c45a19130e36a68)

File: src/utils/convertToSignableToken.ts (hunk 7 above)
Comment (zatlite) [reply src/utils/convertToSignableToken.ts]: [05dab29](https://github.com/immutable/imx-core-sdk/pull/185/commits/05dab29a61dda6416ec0080c8c45a19130e36a68)

File: src/workflows/deposit/depositERC20.ts (hunk 10 above)
Comment (zatlite) [reply src/workflows/deposit/depositERC20.ts]: [05dab29](https://github.com/immutable/imx-core-sdk/pull/185/commits/05dab29a61dda6416ec0080c8c45a19130e36a68)

File: src/workflows/deposit/depositERC20.ts [hunk 15]
```diff
@@ -89,7 +89,7 @@ export async function depositERC20Workflow(
     token_address: deposit.tokenAddress,
   };
 
-  const amount = parseUnits(deposit.amount, 0); // 0 to always use undecimilised value
+  const amount = parseUnits(deposit.amount, 0); // 0 to always use undecimilized value
```
Comment (matthewmuscat) [inline src/workflows/deposit/depositERC20.ts:92]:
```suggestion
  const amount = parseUnits(deposit.amount, 0); // 0 to always use undecimalized value
```

File: src/workflows/withdrawal/completeERC721Withdrawal.ts (hunk 14 above)
Comment (zatlite) [reply src/workflows/withdrawal/completeERC721Withdrawal.ts]: [1df7ea9](https://github.com/immutable/imx-core-sdk/pull/185/commits/1df7ea9f71888977c0a6ffa4a1b00e6851a63039)

File: src/workflows/deposit/depositERC20.ts (hunk 15 above)
Comment (zatlite) [reply src/workflows/deposit/depositERC20.ts]: [6972f32](https://github.com/immutable/imx-core-sdk/pull/185/commits/6972f3238a2df9c5c6a27592c54309bf4df66683)

File: src/workflows/transfers.ts (hunk 13 above)
Comment (zatlite) [reply src/workflows/transfers.ts:84]: Request is an array of `NftTransferDetails` so it is already strictly typed. I'll rename `x` to `nftTransfer`

Comment (zatlite) [reply src/workflows/transfers.ts]: [fc0cb1f](https://github.com/immutable/imx-core-sdk/pull/185/commits/fc0cb1f5944ec32255b0c3f5077e515af50b3b61)

File: src/workflows/registration.ts (hunk 12 above)
Comment (zatlite) [reply src/workflows/registration.ts:63]: Typescript doesn't like that. It wouldn't complain if we did `catch (err: any)` though.

File: src/exportContracts.ts (hunk 6 above)
Comment (zatlite) [reply src/exportContracts.ts:2]: I can't see former names? https://docs.x.immutable.com/sdk-references/core-sdk-ts/1.0.0-beta.2/modules/index.contracts

File: src/workflows/registration.ts (hunk 11 above)
Comment (zatlite) [reply src/workflows/registration.ts:45]: Created DX-1265 for followup.

File: src/workflows/registration.ts (hunk 12 above)
Comment (matthewmuscat) [reply src/workflows/registration.ts:63]: ah makes sense. fair enough

File: src/exportContracts.ts (hunk 6 above)
Comment (matthewmuscat) [reply src/exportContracts.ts:2]: oh nice! looks good to me. @Kastet looks resolved now?

**Review by matthewmuscat — APPROVED** [review]

**Pull request merged**